### PR TITLE
Change vml.h to support sizes greater than 2**32 - 1

### PR DESCRIPTION
**Pull request by cpuhrsch**

### CODE REVIEW

**Review by soumith — CHANGES_REQUESTED** [review]
what happens here if MKL_INT is int64_t? will it silently fail, or does it work fine?

If there's silent failure, can you static_assert

**Comment by gchanan** [comment]
can we add a test?  (how expensive is it?)

**Comment by cpuhrsch** [comment]
@soumith - I think we'd need to worry about size_t, but for int64_t the max value should then cause it always stick to the first branch. 

There's a bunch of stuff left to do here, including benchmarks on the impact of writing these branches (which I'd be surprised if there are any).

A static assert is still a good idea. In general, we should move all includes of mkl.h into a single location and reference our own header. There are some implicit settings one can do by defining variables before the include. This can then also include the static assert, which is a good way to check that PyTorch is only built with the lp64 version (as oppposed to ilp64).

@gchanan - Since this is very specific to MKL VML and it's referenced in a single location, I'm not sure it's worth adding this to the tests we run for every built.

**Comment by gchanan** [comment]
> @gchanan - Since this is very specific to MKL VML and it's referenced in a single location, I'm not sure it's worth adding this to the tests we run for every built.

We need nightlies!

**Comment by cpuhrsch** [comment]
I ran some performance regression tests and the difference seems to be within the noise, so performance wise this should be fine.

File: aten/src/ATen/cpu/vml.h [hunk 1]
```diff
@@ -125,16 +126,42 @@ IMPLEMENT_VML_BUG(trunc)
...
-  inline void v##op(float* out, const float* in, int64_t size) {              \
-    vms##mklop(size, in, out, VML_HA | VML_FTZDAZ_OFF | VML_ERRMODE_IGNORE); \
-  }                                                                           \
-  template <>                                                                 \
-  inline void v##op(double* out, const double* in, int64_t size) {            \
-    vmd##mklop(size, in, out, VML_HA | VML_FTZDAZ_OFF | VML_ERRMODE_IGNORE); \
+// NB: LP64 MKL is the most commonly used and thus we assume it here. That means
+// we need to expect MKL_INT to be of type int, which implies int32_t in most
+// cases.
+static_assert(
```
Comment (soumith) [inline aten/src/ATen/cpu/vml.h]: ship this, but we should add a global setting to allow building / linking against MKL_INT=int64_t as well, it just has to be non-default. Send a patch to do that in a follow-up PR.

**Review by soumith — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
Approved via `ghapprove D14154997`

**Comment by cpuhrsch** [comment]
Thanks. Will merge soon.

**Pull request closed without merging**